fix performance issue due to search in tokens

bagerard opened the pull request:
Fixes #176

I looked into this and the issue was in the following block
```
comment_in_line = any(
    token_type == tokenize.COMMENT
    for token_type, _, _, _, _ in self._tokens)
```

self._tokens is actually containing the tokens from beginning of the file until the physical line yielded by flake8, this means that self._tokens is getting bigger and bigger every time flake8 invokes this plugin on the next lines, thus searching in self._tokens gets more and more expensive for large files as it's inefficient to search in large list by design (the large file is > 12000 lines so that makes a lot of tokens).

I switch the code to process the whole file instead of work line by line, that way it has to search for the comment token only once.

On the long python file attached in #176, this makes the processing time go from 30 sec to just 2 sec :rocket:

sobolevn commented on tests/test_comments.py:
```diff
@@ -120,8 +123,22 @@ def test_incorrect_fixture_whitelist_extend(absolute_path):
         assert b'# typed_property: int = 10' in stdout
+
+
+def test__lines_with_commented_out_code_incorrect_fixture_output(absolute_path):
```
nitpick: `test_lines_...`

sobolevn reviewed (commented):
Hi! Thanks a lot for your work! 

Let's run the CI and check it out. 👍

sobolevn commented on flake8_eradicate.py:
```diff
@@ -25,7 +25,7 @@ class Checker(object):
     options = None
 
-    def __init__(self, tree, filename: str):
+    def __init__(self, tree, filename: str):    # noqa: D107
```
Let's not ignore this one. Let's document what `tree` and `filename` are.
Btw, let's add `ast.AST` type to `tree`

sobolevn reviewed (commented):
Looks like, you've accidentally deleted `pyproject.toml`

bagerard commented:
omg... sorry for that. I'm using pip to install it locally, not poetry so I had to clean those files temporarily...
These manually triggered pipeline are really killing productivity, let's hope github finds another solution to address the cryptomining issues

bagerard commented:
Alright, now it should be good. I've set it up with poetry locally.
I added an additional piece of code to skip the `# -*- coding: utf-8 -*-` line in file (so that it doesn't waste time processing the file if that is the only tokenize.COMMENT found in the file) and I added a test for that

bagerard commented:
Hi, please re-run the pipeline when you have a chance ;)

sobolevn commented:
Done!

sobolevn commented on flake8_eradicate.py:
```diff
@@ -121,19 +125,28 @@ def _contains_commented_out_code(self) -> bool:
-                    self._physical_line,
-                    self._options['aggressive'],
-                ),
+        with open(self.filename) as f:
```
So, my main question right now is: why do we have to construct our tokens here, why can't we use `file_tokens` object from `flake8`?

bagerard replied:
I changed the code to make it use file_tokens in the constructor, I'm not sure it's improving a lot since we still have to open the file to read the physical lines but if that looks better to you, so be it

bagerard commented:
Note that consuming the file_tokens from flake8 made it slightly less trivial/efficient to ignore the file encoding `'# -*- coding:'` comment so I ended up removing that optimization. It was not a major improvement and anyway the `'# -*- coding:'` thing is something from the past so it may be better to avoid polluting the code with that (e.g pyupgrade removes them by default https://github.com/asottile/pyupgrade#-coding--comment)

bagerard commented:
Would you mind running the pipeline on it ?

bagerard commented:
Let me know if there is anything else I should do or if it's fine as it is

sobolevn commented:
@bagerard I will return to this PR sometime soon! Please, stay tuned 🙂

sobolevn commented:
I finally got the time to work on this, sorry for the long wait!

bagerard commented on CHANGELOG.md:
```diff
@@ -3,6 +3,13 @@
+
+### Features
+
+- Imrpoves performance on long files #210
```
typo here :)

sobolevn replied:
Fixed! Thanks!

sobolevn commented:
@bagerard thanks a lot for your work! 👍
